### PR TITLE
[runtime] Add unhandled exception serializer

### PR DESCRIPTION
**Pull request by alexanderkyte**

### CODE REVIEW

**Comment by alexanderkyte** [comment]
After adding a "throw new Exception()" to the below run test:

```akyte@Alexanders-MacBook-Pro-3 ~/mono_master/mcs/class/corlib (nunit_telemetry*) $ ls mono_crash*                
mono_crash.b38e73f1.0.json
akyte@Alexanders-MacBook-Pro-3 ~/mono_master/mcs/class/corlib (nunit_telemetry*) $ make test && export MONO_TEST_TELEMETRY="yes" PATH="/Users/akyte/mono_master/runtime/_tmpinst/bin:/usr/local/Cellar/ninja/1.8.2/bin:/usr/local/opt/ncurses/bin:/Users/akyte/.opam/system/bin:/Library/Frameworks/Mono.framework/Versions/Current/bin/:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/usr/local/share/dotnet:/opt/X11/bin:~/.dotnet/tools:/Library/Frameworks/Mono.framework/Versions/Current/Commands:/Applications/Xamarin Workbooks.app/Contents/SharedSupport/path-bin:" MONO_REGISTRY_PATH="/Users/akyte/.mono/registry" MONO_TESTS_IN_PROGRESS="yes" DBG_RUNTIME_ARGS="" MONO_PATH="./../../class/lib/net_4_x-macos/:./../../class/lib/net_4_x-macos//tests:./../../../external/xunit-binaries:.:$MONO_PATH" ;  /Users/akyte/mono_master/mono/mini/mono-sgen --debug    ./../../class/lib/net_4_x-macos/nunit-lite-console.exe  ./../../class/lib/net_4_x/tests/net_4_x_corlib_test.dll -test:MonoTests.System.ExceptionTest 2>&1
/Users/akyte/Downloads/Xcode.app/Contents/Developer/usr/bin/make test-local
NUnitLite 1.0.0 (.NET 4.5 Debug)
Copyright 2013, Charlie Poole

Runtime Environment -
    OS Version: Unix 17.4.0.0
  Mono Version: 4.0.30319.42000

..........F.
Tests run: 9, Passed: 8, Errors: 1, Failures: 0, Inconclusive: 0
  Not run: 0, Invalid: 0, Ignored: 0, Skipped: 0
Elapsed time: 00:00:00.1200000

Errors and Failures:

1) StacktraceToState (MonoTests.System.ExceptionTest.StacktraceToState)
   System.Exception : #1000
  at (wrapper managed-to-native) System.Reflection.MonoMethod.InternalInvoke(System.Reflection.MonoMethod,object,object[],System.Exception&)
  at System.Reflection.MonoMethod.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x0003b] in /Users/akyte/mono_master/mcs/class/corlib/System.Reflection/MonoMethod.cs:305 

akyte@Alexanders-MacBook-Pro-3 ~/mono_master/mcs/class/corlib (nunit_telemetry*) $ ls mono_crash*
mono_crash.b38e73f1.0.json mono_crash.b38e73f1.1.json
akyte@Alexanders-MacBook-Pro-3 ~/mono_master/mcs/class/corlib (nunit_telemetry*) $ ```

File: NUnitLite-1.0.0/src/framework/Internal/Results/TestResult.cs [hunk 1]
```diff
@@ -412,20 +413,27 @@ public void RecordException(Exception ex)
             if (ex is NUnitException)
                 ex = ex.InnerException;
 
-            if (ex is System.Threading.ThreadAbortException)
+            if (ex is System.Threading.ThreadAbortException) {
```
Comment (marek-safar) [inline NUnitLite-1.0.0/src/framework/Internal/Results/TestResult.cs]: Please undo the unnecessary changes

File: NUnitLite-1.0.0/src/framework/Internal/Results/TestResult.cs [hunk 2]
```diff
@@ -412,20 +413,27 @@ public void RecordException(Exception ex)
...
+            } else if (ex is InconclusiveException) {
               SetResult(ResultState.Inconclusive, ex.Message, StackFilter.Filter(ex.StackTrace));
-            else if (ex is SuccessException)
+            } else if (ex is SuccessException) {
               SetResult(ResultState.Success, ex.Message, StackFilter.Filter(ex.StackTrace));
-            else
-                SetResult(ResultState.Error,
+            } else {
+              MethodInfo write = null;
+              if (Environment.GetEnvironmentVariable ("MONO_TEST_TELEMETRY") != null)
```
Comment (marek-safar) [inline NUnitLite-1.0.0/src/framework/Internal/Results/TestResult.cs]: What is the purpose of this? If you want to test the telemetry it's better to write non-unit test

Comment (alexanderkyte) [reply NUnitLite-1.0.0/src/framework/Internal/Results/TestResult.cs]: The purpose of this is to serialize unhandled exceptions to disk. I assumed this was not a default desired behavior, so I placed the functionality behind an environment variable. This isn't a unit test.

Comment (alexanderkyte) [reply NUnitLite-1.0.0/src/framework/Internal/Results/TestResult.cs]: This is being used on CI to send crashes to sentry

File: NUnitLite-1.0.0/src/framework/Internal/Results/TestResult.cs (hunk 1 above)
Comment (alexanderkyte) [reply NUnitLite-1.0.0/src/framework/Internal/Results/TestResult.cs]: I have been repeatedly told that it is against our style to add braces around an `else` body without also adding it to the `if` body. 

I can undo this change, but I ask that we pick a consistent rule and stick to it.

**Comment by alexanderkyte** [comment]
I have made the requested style change, though not certain that's what we want

File: NUnitLite-1.0.0/src/framework/Internal/Results/TestResult.cs (hunk 2 above)
Comment (marek-safar) [reply NUnitLite-1.0.0/src/framework/Internal/Results/TestResult.cs]: I am not sure just how this will be usefull when there are many failures. Shouldn't there be a way to tweak the random file name?
